### PR TITLE
Update typescript SDK sufficient to be able to make changes to entities.

### PR DESCRIPTION
**Pull request by DanielCollins**

### CODE REVIEW

**Comment by DanielCollins** [comment]
tests like this:

   m = new Merchi();
   p = new m.Product();
   p.id = "1";
   p.name = "new name";
   p.save()

observe: product name changed.

File: typescript/src/entities/domain.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  get id(): number | undefined {
+    return this._id;
+  }
+
+  set id(newId: number | undefined) {
+    this._id = newId;
+    this._isDirty = true;
+  }
+
+  @Domain.property("domain")
```
Comment (s3341458) [inline typescript/src/entities/domain.ts]: nothing serious but a property pretty much needs us to write 10 lines of code, this is pretty sad to me, any better alternative exists?

File: typescript/src/entities/product.test.ts [hunk 2]
```diff
@@ -23,13 +23,30 @@ test('can get and set name', () => {
...
   mockFetch(true, {'product': {'name': testName}}, 200);
   return merchi.Product.get(1).then(product => expect(product.name).toBe(testName));
 });
 
+test('can fetch with explicit session token', () => {
+  const testToken = "YrDwzmh8&QGtAfg9quh(4QfSlE^RPXWl";
+  const merchi = new Merchi(testToken);
+  const testName = 'S7qHUfV_dr5l';
+  mockFetch(true, {'product': {'name': testName}}, 200);
+  (window as any).merchiBackendUri = 'http://override.example.com/';
```
Comment (s3341458) [inline typescript/src/entities/product.test.ts]: I think we have as setup thing and this is not needed any more?

File: typescript/src/entities/product.test.ts [hunk 3]
```diff
@@ -44,6 +61,24 @@ test('can fetch product with category', () => {
...
 
+test('can fetch product with category and explcit session', () => {
+  const testToken = "YrDwzmh8&QGtAfg9quh(4QfSlE^RPXWl";
+  const merchi = new Merchi(testToken);
+  const testName = 'S7qHUfV_dr5l';
+  const categoryName = 'l3VfG#S+';
+  const categoryData = {'name': categoryName};
+  mockFetch(true, {'product': {'name': testName,
+                               'categories': [categoryData]}}, 200);
+  (window as any).merchiBackendUri = 'http://override.example.com/';
```
Comment (s3341458) [inline typescript/src/entities/product.test.ts]: as above

File: typescript/src/entities/product.test.ts [hunk 4]
```diff
@@ -84,3 +119,35 @@ test('can list products from server with category', () => {
...
+  p.save();
+  (global as any).fetch = jest.fn().mockImplementation((url, data) => {
+    expect(data.method).toBe("PATCH");
+    const correct =  [['products-0-name', 'pHyz7ZucK#'],
+      ['products-0-categories-0-name', '8&OaUsDgJ$ev3FYZ3'],
+      ['products-0-categories-count', '1'],
+      ['products-0-domain-0-domain', '3onrb6o4'],
+      ['products-0-domain-count', '1'],
+      ['products-count', '1']];
+    expect(Array.from(data.body.entries())).toEqual(correct);
```
Comment (s3341458) [inline typescript/src/entities/product.test.ts]: it is magic here to me, how did you guarantee Array.from(data.body.entries()) and "correct" has the same order

Another good idea actually bring by @Amtron that we should use a nested json as correct and serialise it instead directly write it as flat out json (hope I explained properly)

Whether the serialization function is correct is a independent thing that needs to test.

this makes the code more readable and scalable, if the protocol changed we do not need to fix all testcases only need to adjust the testcase for serialization function

Comment (s3341458) [reply typescript/src/entities/product.test.ts]: serialization function refers the toFormData you done blow

File: typescript/src/merchi.ts [hunk 5]
```diff
@@ -20,9 +24,17 @@ export class Merchi {
...
 
   public Category: typeof Category
+  public Domain: typeof Domain;
   public Product: typeof Product;
 
-  constructor() {
+  public sessionToken?: string;
+
+  constructor(sessionToken?: string) {
+    if (sessionToken) {
```
Comment (s3341458) [inline typescript/src/merchi.ts]: I was struggled when look at it, I think you might be struggled to make this decision as well.
I think it is fine for now

File: typescript/src/entity.ts [hunk 6]
```diff
@@ -198,4 +219,71 @@ export class Entity {
...
     }
   }
+
+  protected forEachProperty = (fn: (i: PropertyInfo) => void) => {
+    for (const info of this.propertiesMap.values()) {
+      fn(info);
+    }
+  }
+
+  protected toFormData = (options?: SerialiseOptions): FormData => {
```
Comment (s3341458) [inline typescript/src/entity.ts]: toFormData worth a independent testcase

File: typescript/src/entities/domain.ts (hunk 1 above)
Comment (DanielCollins) [reply typescript/src/entities/domain.ts]: already discussed previously: https://github.com/merchisdk/sdk/pull/181

File: typescript/src/entities/product.test.ts (hunk 2 above)
Comment (DanielCollins) [reply typescript/src/entities/product.test.ts]: correct

File: typescript/src/entities/product.test.ts (hunk 4 above)
Comment (DanielCollins) [reply typescript/src/entities/product.test.ts]: > how did you guarantee Array.from(data.body.entries()) and "correct" has the same order

the order is of course not important to the protocol currently (though there is actually movement to change that in some cases).

`toFormData` however is detirministic, and therefore always traverses the graph in the same order, and therefore always outputs in the same order. MDN doesn't mention anything about the ordering of `FormData.entries()`, but it was be surprising to me if they used a non-deterministic iterator.

Comment (DanielCollins) [reply typescript/src/entities/product.test.ts]: > that we should use a nested json as correct and serialise it instead directly write it as flat out json 

what's the supposed benefit of splitting the transformation over two steps?

Ultimately these entity objects are just javascript objects the same as JSON obejcts would be. They just contain some extra information, too. Why spend an extra step removing that extra information, when the final step would presumably be the same regardless?

File: typescript/src/merchi.ts (hunk 5 above)
Comment (DanielCollins) [reply typescript/src/merchi.ts]: looking up cookies using document.cookie is a bit of a hack, in particular, i don't know how well that's going to work on node or a another non-browser environment.

but, line 33 that you pointed out, i thought it was quite clear. straight forward check for whether or not the argument was provided.... what part don't you like?

**Comment by DanielCollins** [comment]
i have updated this branch to include a seperate test for `toFormData`

File: typescript/src/entities/product.test.ts [hunk 7]
```diff
@@ -96,3 +143,50 @@ test('can list products from server with category', () => {
...
+  p.domain = d;
+  c1.products = [p];
+  c1.save();
+  d.domain = "3onrb6o4";
+  p.name = "pHyz7ZucK#";
+  c2.name = "8&OaUsDgJ$ev3FYZ3";
+  p.save();
+  const fetch = mockFetch(true, {}, 200);
+  c1.save();
+  const correct =  [['products-0-name', 'pHyz7ZucK#'],
```
Comment (s3341458) [inline typescript/src/entities/product.test.ts]: I personally think make correct  =  p.toFormData() due avoid duplicate testing

as p.toFormData has been tested in another testcase

Comment (s3341458) [reply typescript/src/entities/product.test.ts]: I will approve it and leave this as suggestion

Comment (DanielCollins) [reply typescript/src/entities/product.test.ts]: that's not a bad idea. i'll think about it for next time

**Pull request merged**